manticore/core/cpu/abstractcpu.py

```diff
@@ -479,89 +554,98 @@ def read_bytes(self, where, size):
+            # Read Instruction from memory
             for i in xrange(0, self.max_instr_width):
-                c = self.memory[pc+i]
-                if issymbolic(c):
-                    assert isinstance(c, BitVec) and  c.size == 8
-                    if isinstance(c, Constant):
-                        c = chr(c.value)
-                    else:
-                        logger.error('Concretize executable memory %r %r', c, text )
-                        break
-                assert isinstance(c, str)
-                text += c
-        except MemoryException:
+                text += _read_concrete_byte(pc+i)
+        except MemoryException as mem_except:
+            #lets not raise this exception in case the instruction
+            #was fully read befor the page
             pass
-        
+
+        #Pad potentially incomplete intruction with zeroes
         code = text.ljust(self.max_instr_width, '\x00')
-        instruction = next(self._md.disasm(code, pc))
 
-        #PC points to symbolic memory 
-        if instruction.size > len(text):
-            logger.info("Trying to execute instructions from invalid memory")
-            raise InvalidPCException(pc)
+        #decode the instructtion from code 
+        try:
+            instruction = next(self._md.disasm(code, pc))
+        except Exception as e:
+            #Fixme/Caveat  will raise
+            raise DecodeException(pc, code) 
 
+        #Check that the decoded intruction is contained in executable memory
         if not self.memory.access_ok(slice(pc, pc+instruction.size), 'x'):
             logger.info("Trying to execute instructions from non-executable memory")
-            raise InvalidPCException(pc)
+            raise InvalidMemoryAccess(pc, 'x')
 
         instruction.operands = self._wrap_operands(instruction.operands)
 
         self._instruction_cache[pc] = instruction
-        self.instruction = instruction
-
+        return instruction
 
     #######################################
     # Execute
-    @abstractmethod
     def canonicalize_instruction_name(self, instruction):
         '''
         Get the semantic name of an instruction. 
         '''
-        pass
+        raise NotImplemented
 
     def execute(self):
         '''
         Decode, and execute one instruction pointed by register PC
         '''
 
-        # Decode the instruction if it wasn't explicitly decoded
-        if self.instruction is None or self.instruction.address != self.PC:
-            self.decode_instruction(self.PC)
+        if issymbolic(self.PC):
+            raise ConcretizeRegister(self, 'PC', policy='ALL')
 
-        instruction = self.instruction
+        if not self.memory.access_ok(self.PC,'x'):
+            raise InvalidMemoryAccess(self.PC, 'x')
+
+        #broadcast event
+        self.will_decode_instruction()
+
+        instruction = self.decode_instruction(self.PC)
+
+        #broadcast event
+        self.will_execute_instruction(instruction)
```

Comment:
This is how the callback for this signal will look like at manticore level. 
   def _execute_instruction_callback(self, state, platform, cpu, instruction):
        logger.info("exe\n")

Instruction will be available as an argument and as cpu.intruction both.